develop guide: add sample applications (nextjs and lambda) (#14532)
pingcap/docs: develop/dev-guide-sample-application-aws-lambda.md

Reviewer: preview: https://pingcap-docs-website-scaffold-git-preview-ping-224a19-ran-huang.vercel.app/tidb/dev/dev-guide-sample-application-aws-lambda

TiDB is a MySQL-compatible database.

[mysql2](https://www.npmjs.com/package/mysql2) is one of the most popular open-source Node.js drivers.

[AWS Lambda Function](https://aws.amazon.com/lambda/) is a serverless computing service provided by AWS.

This document shows you how to use TiDB and mysql2 to build a simple CRUD application.

## Prerequisites

- [Node.js **18**](https://nodejs.org/en/download/) or later.
- [Git](https://git-scm.com/downloads).
- TiDB cluster. If you don't have a TiDB cluster, you can create one by following the steps below:
    - (Recommended) Refer to [Create a TiDB Serverless cluster](/develop/dev-guide-build-cluster-in-cloud.md#step-1-create-a-tidb-serverless-cluster) to create your own TiDB Cloud cluster.
    - Refer to [Deploy a local test TiDB cluster](/quick-start-with-tidb.md#deploy-a-local-test-cluster) or [Deploy a formal TiDB cluster](/production-deployment-using-tiup.md) to create a local cluster.
- AWS account. If you don't have an AWS account, you can create one by following the steps below:
    - Refer to [Create an AWS Lambda Function](https://docs.aws.amazon.com/lambda/latest/dg/getting-started.html) to create an AWS Lambda Function.

> 

Reviewer 2: To complete this guide, you need:

- [Node.js](https://nodejs.org/en/download/)  18 or later
- [Git](https://git-scm.com/downloads)
- A TiDB cluster running
- [An AWS account](https://portal.aws.amazon.com/billing/signup)
- An AWS user can access the Lambda function

**If you don't have a TiDB cluster yet, please create one with one of the following methods:**

- (Recommended) Refer to [Create a TiDB Serverless cluster](/develop/dev-guide-build-cluster-in-cloud.md#step-1-create-a-tidb-serverless-cluster) to create your own TiDB Cloud cluster.
- Refer to [Deploy a local test TiDB cluster](/quick-start-with-tidb.md#deploy-a-local-test-cluster) or [Deploy a formal TiDB cluster](/production-deployment-using-tiup.md) to create a local cluster.

**If you don't have an AWS account and user, you can create them by following the steps  in the [Getting started with Lambda](https://docs.aws.amazon.com/lambda/latest/dg/getting-started.html) guide.**

## Run the sample application

This section shows you how to run the sample application code and connect to TiDB.

### Step 1: Clone the sample application repository to your local machine

```bash

> git clone git@github.com:tidb-samples/tidb-aws-lambda-quickstart.git

Reviewer 2: ```suggestion
git clone git@github.com:tidb-samples/tidb-aws-lambda-quickstart.git
cd tidb-aws-lambda-quickstart
```

```

### Step 2: Install dependencies

```bash

> cd tidb-aws-lambda-quickstart

Reviewer 2: ```suggestion
```

   ```json
   {
     "Parameters": {
       "TIDB_HOST": "{gateway-region}.aws.tidbcloud.com",
       "TIDB_PORT": "4000",
       "TIDB_USER": "{prefix}.root",
       "TIDB_PASSWORD": "{password}"
     }
   }
   ```

   Replace the placeholders in `{}` with the values obtained in the **Connect** window.

5. [Configure Step.4 environment variables](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) in Lambda Function.

</div>

<div label="Self-hosted TiDB">

1. Copy and paste the corresponding connection string. The following is an example:

   ```json
   {
     "Parameters": {
       "TIDB_HOST": "{tidb_server_host}",
       "TIDB_PORT": "4000",
       "TIDB_USER": "root",
       "TIDB_PASSWORD": "{password}"
     }
   }
   ```

   Replace the placeholders in `{}` with the values obtained in the **Connect** window.

2. [Configure Step.1 environment variables](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) in Lambda Function..

</div>

</SimpleTab>

### Step 4: Run the sample application locally

1. Install the [AWS SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html).

> 2. Run the following command to start the application:

Reviewer 2: ```suggestion
2. Run the following command to invoke the sample Lambda function:
```

    ```bash
    sam local invoke --env-vars env.json -e event.json "tidbHelloWorldFunction"
    ```

3. Check the output in the terminal. If the output is similar to the following, the connection is successful:

    ```bash
    {"statusCode":200,"body":"{\"results\":[{\"Hello World\":\"Hello World\"}]}"}
    ```

> ## Key Code Snippets

Reviewer 2: ```suggestion
## Example codes
```

Complete code snippets are available in the [tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) GitHub repository.

You can refer to the following key code snippets to complete your application development.

### Connect to TiDB

```typescript
import mysql from 'mysql2';

const pool = mysql.createPool({

>   host,
>   port,
>   user,
>   password,
>   database,

Reviewer 2: The declaration of these variables is not mentioned in the code:

```suggestion
  host,
  port,
  user,
  password,
  database,
```

  ssl: {
    minVersion: 'TLSv1.2',
    rejectUnauthorized: true,
  },

>   waitForConnections: true,
>   connectionLimit: 1,
>   maxIdle: 1, // max idle connections, the default value is the same as `connectionLimit`
>   idleTimeout: 60000, // idle connections timeout, in milliseconds, the default value 60000
>   queueLimit: 0,
>   enableKeepAlive: true,
>   keepAliveInitialDelay: 0,

Reviewer 2: Keep it simple, delete parameters with default values, It may make beginners feel confused.

```suggestion
  connectionLimit: 1,
  maxIdle: 1, // max idle connections, the default value is the same as `connectionLimit`
  enableKeepAlive: true
```

});
```

### Insert data

```typescript
const player = ['1', 10, 500];
await pool.execute(
  `INSERT INTO player (id, goods, coins) VALUES (${player[0]}, ${player[1]}, ${player[2]})`
);
```

Refer to [Insert data](/develop/dev-guide-insert-data.md) for more information.

### Query data

```typescript
const results = await pool.execute('SELECT count(*) FROM player');
console.log(results);
```

Refer to [Query data](/develop/dev-guide-get-data-from-single-table.md) for more information.

### Update data

```typescript
const player = ['1', 10, 500];
await pool.execute(
  `UPDATE player SET goods = goods + ${play[0]}, coins = coins + ${play[1]} WHERE id = ${play[2]}`
);
```

Refer to [Update data](/develop/dev-guide-update-data.md) for more information.

### Delete data

```typescript
await pool.execute('DELETE FROM player WHERE id = 1');
```

Refer to [Delete data](/develop/dev-guide-delete-data.md) for more information.

## Considerations

> - Complete code snippets and how to run them, see [tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) GitHub repository.

Reviewer 2: Remove it or add it to the top of this document like:

<img width="1056" alt="image" src="https://github.com/pingcap/docs/assets/5086433/d7e7a190-961e-48f3-ae3c-fdb0d952ae27">

> - The driver is not highly encapsulated, so you will see a lot of SQL statements in the program. Unlike ORM, because there is no data object, the query object of `mysql2` is represented by a object. Although the Node.js driver is convenient, it still requires manual control of the transaction characteristics because it cannot shield the underlying implementation. If there are no scenarios that must use SQL, it is still recommended to use ORM to write programs. This can reduce the coupling of the program.

Reviewer 2: Can we avoid long paragraphs in list items? Shotern it may be better.

Reference: https://www.prisma.io/docs/about/prisma-docs/style-guide/writing-style#simplify

```suggestion
- The driver is not highly encapsulated, so you will see a lot of SQL statements in the program. Unlike ORM, because there is no data object, the query object of `mysql2` is represented by a object. Although the Node.js driver is convenient, it still requires manual control of the transaction characteristics because it cannot shield the underlying implementation. If there are no scenarios that must use SQL, it is still recommended to use ORM to write programs. This can reduce the coupling of the program.
```

> - For more information about how to use mysql2, see [mysql2 official documentation](https://github.com/sidorares/node-mysql2).
> - For more information about how to use AWS Lambda Function with TiDB Cloud, see [Designing High-Performance Applications Using Serverless TiDB Cloud and AWS Lambda](https://aws.amazon.com/blogs/apn/designing-high-performance-applications-using-serverless-tidb-cloud-and-aws-lambda/).

Reviewer 2: I think It is better to place these resource links in the **What's next** / **Resources** section instead of **Considerations** section.

```suggestion
```

## What's next

> - You can continue to read the developer documentation to learn more about TiDB. For example: [Insert data](/develop/dev-guide-insert-data.md), [Update data](/develop/dev-guide-update-data.md), [Delete data](/develop/dev-guide-delete-data.md), [Get data from single table](/develop/dev-guide-get-data-from-single-table.md), [Transaction](/develop/dev-guide-transaction-overview.md), [SQL performance optimization](/develop/dev-guide-optimize-sql-overview.md), etc.

Reviewer 2: Make a list instead of long statement.

1. Install the [AWS SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html).

2. Run the following command to invoke the sample Lambda function:

    ```bash
    sam local invoke --env-vars env.json -e event.json "tidbHelloWorldFunction"
    ```

3. Check the output in the terminal. If the output is similar to the following, the connection is successful:

    ```bash
    {"statusCode":200,"body":"{\"results\":[{\"Hello World\":\"Hello World\"}]}"}
    ```

## Example codes

Complete code snippets are available in the [tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) GitHub repository.

You can refer to the following key code snippets to complete your application development.

### Connect to TiDB

```typescript
import mysql from 'mysql2';

const pool = mysql.createPool({
  host, // TiDB host, for example: {gateway-region}.aws.tidbcloud.com
  port, // TiDB port, default: 4000
  user, // TiDB user, for example: {prefix}.root
  password, // TiDB password
  database, // TiDB database name, default: test
  ssl: {
    minVersion: 'TLSv1.2',
    rejectUnauthorized: true,
  },

>   connectionLimit: 1,

Reviewer 2: Maybe we should briefly explain the reason why the `connectionLimit`,`maxIdle` is set to `1`.

- https://www.cockroachlabs.com/docs/v23.1/serverless-function-best-practices#use-connection-pools
- https://www.prisma.io/docs/guides/deployment/serverless/deploy-to-aws-lambda#connection-pooling
- https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#without-an-external-connection-pooler

  maxIdle: 1, // max idle connections, the default value is the same as `connectionLimit`
  enableKeepAlive: true
});

> ```

Reviewer 2: For the Serverless environment, it is important to place the declaration of the connection pool outside of the handler so that it can be reused in different invocations, rather than always establishing a new connection every invocations.

```javascript
let pool;

exports.handler = async (context) => {
  // some code ...
}
```

Reference: https://www.cockroachlabs.com/docs/v23.1/serverless-function-best-practices#persist-connection-pools-across-function-invocations

---

> title: Build a CRUD Application with TiDB and AWS Lambda Function

Author: ```suggestion
title: Connect to TiDB by using mysql2 in AWS Lambda Function
```

> summary: This article describes how to connect TiDB using mysql2 in AWS Lambda Function and provides a simple example code snippet.

Reviewer: ```suggestion
summary: This article describes how to build a CRUD application using TiDB and mysql2 in AWS Lambda Function and provides a simple example code snippet.
```

---

<!-- markdownlint-disable MD024 -->
<!-- markdownlint-disable MD029 -->

> # Build a CRUD Application with TiDB and mysql2 in AWS Lambda Function
> 
> This guide demonstrates how to use TiDB, a MySQL-compatible database, and [mysql2](https://www.npmjs.com/package/mysql2), a popular open-source Node.js driver, in an [AWS Lambda Function](https://aws.amazon.com/lambda/) to build a simple CRUD application.

Author: ```suggestion
# Connect to TiDB by using mysql2 in AWS Lambda Function

This guide demonstrates how to use TiDB, a MySQL-compatible database, and [mysql2](https://www.npmjs.com/package/mysql2), a popular open-source Node.js driver, in an [AWS Lambda Function](https://aws.amazon.com/lambda/) to build a simple CRUD application.

TiDB is a MySQL-compatible database, [AWS Lambda Function](https://aws.amazon.com/lambda/) is a compute service, and [mysql2](https://github.com/sidorares/node-mysql2) is a popular open-source driver for Node.js.

In this tutorial, you can learn how to use TiDB and mysql2 in AWS Lambda Function to accomplish the following tasks:

- Set up your environment.
- Connect to your TiDB cluster using mysql2.
- Build and run your application. Optionally, you can find sample code snippets for basic CRUD operations.

> **Note:**
>
> This tutorial works with TiDB Serverless and TiDB Self-Hosted.
```

## Prerequisites

Ensure you have the following installed and set up:

- [Node.js **18**](https://nodejs.org/en/download/) or later.
- [Git](https://git-scm.com/downloads).
- A running TiDB cluster.
- An [AWS account](https://repost.aws/knowledge-center/create-and-activate-aws-account).
- An AWS user with access to the Lambda function.

If you don't have a TiDB cluster yet, you can create one using one of the following methods:

- (Recommended) Refer to [Create a TiDB Serverless cluster](/develop/dev-guide-build-cluster-in-cloud.md) to create your own TiDB Cloud cluster.
- Refer to [Deploy a local test TiDB cluster](/quick-start-with-tidb.md) or [Deploy a formal TiDB cluster](/production-deployment-using-tiup.md) to create a local cluster.

> If you don't have an AWS account and user, you can create them by following the steps in the [Getting started with Lambda](https://docs.aws.amazon.com/lambda/latest/dg/getting-started.html) guide.

Reviewer: ```suggestion
If you don't have an AWS account or a user, you can create them by following the steps in the [Getting Started with Lambda](https://docs.aws.amazon.com/lambda/latest/dg/getting-started.html) guide.
```

  });
}

export async function handler(event: any) {
  if (!pool) {
    connect();
  }
  const results = await pool.execute('SELECT "Hello World"');
  return {
    statusCode: 200,
    body: JSON.stringify({ results }),
  };
}
```

### Insert data

```typescript
const player = ['1', 10, 500];
await pool.execute(
  `INSERT INTO player (id, goods, coins) VALUES (${player[0]}, ${player[1]}, ${player[2]})`
);
```

Refer to [Insert data](/develop/dev-guide-insert-data.md) for more information.

### Query data

```typescript
const [rows] = await pool.execute('SELECT count(*) AS cnt FROM player');
console.log(rows[0]['cnt']);
```

Refer to [Query data](/develop/dev-guide-get-data-from-single-table.md) for more information.

### Update data

```typescript
const player = ['1', 10, 500];
await pool.execute(
  `UPDATE player SET goods = goods + ${play[0]}, coins = coins + ${play[1]} WHERE id = ${play[2]}`
);
```

Refer to [Update data](/develop/dev-guide-update-data.md) for more information.

### Delete data

```typescript
await pool.execute('DELETE FROM player WHERE id = 1');
```

Refer to [Delete data](/develop/dev-guide-delete-data.md) for more information.

## Considerations

> - It's recommended that using ORM frameworks to improve development efficiency in scenarios without a lot of complex SQL, such as [Sequelize](https://sequelize.org/), [Prisma](https://www.prisma.io/) and [TypeORM](https://typeorm.io/).

Reviewer: ```suggestion
- It is recommended to use ORM frameworks like [Sequelize](https://sequelize.org/), [Prisma](https://www.prisma.io/), and [TypeORM](https://typeorm.io/) to improve development efficiency in scenarios that do not involve a lot of complex SQL.

> ## Considerations
> 
> - It is recommended to use ORM frameworks like [Sequelize](https://sequelize.org/), [Prisma](https://www.prisma.io/), and [TypeORM](https://typeorm.io/) to improve development efficiency in scenarios that do not involve a lot of complex SQL.
> 
> ## What's next
> 
> To deepen your understanding of TiDB, you can explore the developer documentation, which covers various topics such as:
> 
> - [Insert data](/develop/dev-guide-insert-data.md)
> - [Update data](/develop/dev-guide-update-data.md)
> - [Delete data](/develop/dev-guide-delete-data.md)
> - [Get data from single table](/develop/dev-guide-get-data-from-single-table.md)
> - [Transaction](/develop/dev-guide-transaction-overview.md)
> - [SQL performance optimization](/develop/dev-guide-optimize-sql-overview.md)

Author: ```suggestion
## Useful notes

### Using driver or ORM framework?

The Node.js driver provides low-level access to the database, but it requires the developers to:

- Manually establish and release database connections.
- Manually manage database transactions.
- Manually map data rows to data objects.

Unless you need to write complex SQL statements, it is recommended to use [ORM](https://en.wikipedia.org/w/index.php?title=Object-relational_mapping) framework for development, such as [Sequelize](https://sequelize.org/), [Prisma](https://www.prisma.io/), and [TypeORM](https://typeorm.io/). It can help you:

- Reduce [boilerplate code](https://en.wikipedia.org/wiki/Boilerplate_code) for managing connections and transactions.
- Manipulate data with data objects instead of a number of SQL statements.

## Next steps

- Learn more usage of `mysql2` from [the documentation of `mysql2`](https://github.com/sidorares/node-mysql2/tree/master/documentation/en).
- Learn more usage of `AWS Lambda` from [the AWS developer guide of `Lambda`](https://docs.aws.amazon.com/lambda/latest/dg/welcome.html).
- Learn the best practices for TiDB application development with the chapters in the [Developer guide](/develop/dev-guide-overview.md), such as [Insert data](/develop/dev-guide-insert-data.md), [Update data](/develop/dev-guide-update-data.md), [Delete data](/develop/dev-guide-delete-data.md), [Single table reading](/develop/dev-guide-get-data-from-single-table.md), [Transactions](/develop/dev-guide-transaction-overview.md), and [SQL performance optimization](/develop/dev-guide-optimize-sql-overview.md).
- Learn through the professional [TiDB developer courses](https://www.pingcap.com/education/) and earn [TiDB certifications](https://www.pingcap.com/education/certification/) after passing the exam.

## Need help?

Ask questions on the [Discord](https://discord.gg/vYU9h56kAX), or [create a support ticket](https://support.pingcap.com/).
```

const player = ['1', 10, 500];
await pool.execute(
  `UPDATE player SET goods = goods + ${play[0]}, coins = coins + ${play[1]} WHERE id = ${play[2]}`
);
```

Refer to [Update data](/develop/dev-guide-update-data.md) for more information.

### Delete data

```typescript
await pool.execute('DELETE FROM player WHERE id = 1');
```

Refer to [Delete data](/develop/dev-guide-delete-data.md) for more information.

## Useful notes

- Using [connection pools](https://github.com/sidorares/node-mysql2#using-connection-pools) to manage database connections, which can reduce the performance overhead caused by frequently establishing/destroying connections.
- Using [prepared statements](https://github.com/sidorares/node-mysql2#using-prepared-statements) to avoid SQL injection.
- Using ORM frameworks to improve development efficiency in scenarios without a number of complex SQL statements, such as: [Sequelize](https://sequelize.org/), [TypeORM](https://typeorm.io/), and [Prisma](https://www.prisma.io/).
- It's recommended to [use AWS Lambda with API Gateway](https://docs.aws.amazon.com/lambda/latest/dg/services-apigateway.html) to build a RESTful API for your application.

## Next steps

> - For more details on how to use TiDB in AWS Lambda Function, see our [TiDB-Lambda-integration/aws-lambda-bookstore Demo](https://github.com/pingcap/TiDB-Lambda-integration/blob/main/aws-lambda-bookstore/README.md). And you can use AWS API Gateway to build a RESTful API for your application.

Reviewer: ```suggestion
- For more details on how to use TiDB in AWS Lambda Function, see our [TiDB-Lambda-integration/aws-lambda-bookstore Demo](https://github.com/pingcap/TiDB-Lambda-integration/blob/main/aws-lambda-bookstore/README.md). You can also use AWS API Gateway to build a RESTful API for your application.
```

- Learn more usage of `mysql2` from [the documentation of `mysql2`](https://github.com/sidorares/node-mysql2/tree/master/documentation/en).

> - Learn more usage of `AWS Lambda` from [the AWS developer guide of `Lambda`](https://docs.aws.amazon.com/lambda/latest/dg/welcome.html).

Reviewer: ```suggestion
- Learn more usage of AWS Lambda from [the AWS developer guide of `Lambda`](https://docs.aws.amazon.com/lambda/latest/dg/welcome.html).
```

        "TIDB_PORT": "4000",
        "TIDB_USER": "{prefix}.root",
        "TIDB_PASSWORD": "{password}"
      }
    }
    ```

    Replace the placeholders in `{}` with the values obtained in the connection dialog.

5. [Copy and configure the corresponding connection string](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) in Lambda Function.

    ![quickstart-lambda-env](/media/develop/quickstart-lambda-env.png)

</div>

<div label="TiDB Self-Hosted">

1. Copy and paste the corresponding connection string into `env.json`. The following is an example:

   ```json
   {
     "Parameters": {
       "TIDB_HOST": "{tidb_server_host}",
       "TIDB_PORT": "4000",
       "TIDB_USER": "root",
       "TIDB_PASSWORD": "{password}"
     }
   }
   ```

   Replace the placeholders in `{}` with the values obtained in the **Connect** window.

2. [Copy and configure the corresponding connection string](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) in Lambda Function.

  ![quickstart-lambda-env](/media/develop/quickstart-lambda-env.png)

</div>

</SimpleTab>

### Step 4: Run the sample application locally

> 1. Install the [AWS SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html).

Reviewer 2: Should be Prerequisites?

### Update data

```typescript
const player = ['1', 10, 500];
await pool.execute(
  `UPDATE player SET goods = goods + ${play[0]}, coins = coins + ${play[1]} WHERE id = ${play[2]}`
);
```

Refer to [Update data](/develop/dev-guide-update-data.md) for more information.

### Delete data

```typescript
await pool.execute('DELETE FROM player WHERE id = 1');
```

Refer to [Delete data](/develop/dev-guide-delete-data.md) for more information.

## Useful notes

### Using driver or ORM framework?

The Node.js driver provides low-level access to the database, but it requires the developers to:

- Manually establish and release database connections.
- Manually manage database transactions.
- Manually map data rows to data objects.

Unless you need to write complex SQL statements, it is recommended to use [ORM](https://en.wikipedia.org/w/index.php?title=Object-relational_mapping) framework for development, such as [Sequelize](https://sequelize.org/), [Prisma](https://www.prisma.io/), and [TypeORM](https://typeorm.io/). It can help you:

- Reduce [boilerplate code](https://en.wikipedia.org/wiki/Boilerplate_code) for managing connections and transactions.

> - Manipulate data with data objects instead of a number of SQL statements.

Reviewer 2: Provide some more useful notes about lambda functions/next.js instead of repeating `Useful notes` from the Python tutorial

---
title: Connect to TiDB by using mysql2 in AWS Lambda Function
summary: This article describes how to build a CRUD application using TiDB and mysql2 in AWS Lambda Function and provides a simple example code snippet.
---

# Connect to TiDB by using mysql2 in AWS Lambda Function

TiDB is a MySQL-compatible database, [AWS Lambda Function](https://aws.amazon.com/lambda/) is a compute service, and [mysql2](https://github.com/sidorares/node-mysql2) is a popular open-source driver for Node.js.

In this tutorial, you can learn how to use TiDB and mysql2 in AWS Lambda Function to accomplish the following tasks:

- Set up your environment.
- Connect to your TiDB cluster using mysql2.
- Build and run your application. Optionally, you can find sample code snippets for basic CRUD operations.

> > **Note:**

Reviewer 2: ```suggestion
> **Note**
```

GitHub defines a set of markdown syntax for considerations, perhaps we can refer to it:

https://github.com/orgs/community/discussions/16925

- Follow [Deploy a local test TiDB cluster](/quick-start-with-tidb.md#deploy-a-local-test-cluster) or [Deploy a production TiDB cluster](/production-deployment-using-tiup.md) to create a local cluster.

</CustomContent>
<CustomContent platform="tidb-cloud">

**If you don't have a TiDB cluster, you can create one as follows:**

- (Recommended) Follow [Creating a TiDB Serverless cluster](/develop/dev-guide-build-cluster-in-cloud.md) to create your own TiDB Cloud cluster.
- Follow [Deploy a local test TiDB cluster](https://docs.pingcap.com/tidb/stable/quick-start-with-tidb#deploy-a-local-test-cluster) or [Deploy a production TiDB cluster](https://docs.pingcap.com/tidb/stable/production-deployment-using-tiup) to create a local cluster.

</CustomContent>

If you don't have an AWS account or a user, you can create them by following the steps in the [Getting Started with Lambda](https://docs.aws.amazon.com/lambda/latest/dg/getting-started.html) guide.

## Run the sample application

This section demonstrates how to run the sample application code and connect to TiDB.

> **Note**
>

> > For complete code snippets and running instructions, refer to the[tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) GitHub repository.

Reviewer 2: ```suggestion
> For complete code snippets and running instructions, refer to the [tidb-samples/tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) GitHub repository.
```

```bash
git clone git@github.com:tidb-samples/tidb-aws-lambda-quickstart.git
cd tidb-aws-lambda-quickstart
```

### Step 2: Install dependencies

Run the following command to install the required packages (including `mysql2`) for the sample app:

```bash
npm install
```

### Step 3: Configure connection information

The method to connect to the TiDB cluster varies based on your deployment method.

<SimpleTab>

<div label="TiDB Serverless">

1. In the TiDB Cloud [Clusters](https://tidbcloud.com/console/clusters) page, select your TiDB Serverless cluster and go to the **Overview** page. Click **Connect** in the upper right corner.

2. In the connection dialog, select `General` from the **Connect With** dropdown and keep the default setting of the **Endpoint Type** as `Public`.

    > **Note**
    >
    > In Node.js applications, you don't have to provide an SSL CA certificate, because Node.js uses the built-in [Mozilla CA certificate](https://wiki.mozilla.org/CA/Included_Certificates) by default when establishing the TLS (SSL) connection.

3. If you have not set a password yet, click **Create password** to generate a random password.

>     <Tip>

Reviewer 2: Better to using markdown syntax.

   {
     "Parameters": {
       "TIDB_HOST": "{tidb_server_host}",
       "TIDB_PORT": "4000",
       "TIDB_USER": "root",
       "TIDB_PASSWORD": "{password}"
     }
   }
   ```

   Replace the placeholders in `{}` with the values obtained in the **Connect** window.

2. [Copy and configure the corresponding connection string](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) in Lambda Function.

  ![quickstart-lambda-env](/media/develop/quickstart-lambda-env.png)

</div>

</SimpleTab>

### Step 4: Run the sample application locally

1. (Prerequisite) Install the [AWS SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html).

2. Invoke the sample Lambda function:

    ```bash
    sam local invoke --env-vars env.json -e events/event.json "tidbHelloWorldFunction"
    ```

3. Check the output in the terminal. If the output is similar to the following, the connection is successful:

    ```bash
    {"statusCode":200,"body":"{\"results\":[{\"Hello World\":\"Hello World\"}]}"}
    ```

## Sample code snippets

You can refer to the following sample code snippets to complete your own application development.

> For complete sample code and how to run it, check out the [tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) repository.

Reviewer 2: ```suggestion
For complete sample code and how to run it, check out the [tidb-samples/tidb-aws-lambda-quickstart](https://github.com/tidb-samples/tidb-aws-lambda-quickstart) repository.
```

---
title: Connect to TiDB with mysql2 in AWS Lambda Function
summary: This article describes how to build a CRUD application using TiDB and mysql2 in AWS Lambda Function and provides a simple example code snippet.
---

# Connect to TiDB with mysql2 in AWS Lambda Function

TiDB is a MySQL-compatible database, [AWS Lambda Function](https://aws.amazon.com/lambda/) is a compute service, and [mysql2](https://github.com/sidorares/node-mysql2) is a popular open-source driver for Node.js.

In this tutorial, you can learn how to use TiDB and mysql2 in AWS Lambda Function to accomplish the following tasks:

- Set up your environment.
- Connect to your TiDB cluster using mysql2.

> - Build and run your application. Optionally, you can find sample code snippets for basic CRUD operations.

Reviewer 2: ```suggestion
- Build and run your application. Optionally, you can find [sample code snippets](#sample-code-snippets) for basic CRUD operations.
- Deploy your AWS Lambda Function.
```

> **Note**
>
> This tutorial works with TiDB Serverless and TiDB Self-Hosted.

## Prerequisites

To complete this tutorial, you need:

- [Node.js **18**](https://nodejs.org/en/download/) or later.
- [Git](https://git-scm.com/downloads).
- A TiDB cluster.
- An [AWS account](https://repost.aws/knowledge-center/create-and-activate-aws-account).

> - An AWS user with access to the Lambda function.

Reviewer 2: If we need to use `sam deploy`, we may also need permissions involving `CloudFormation`, `S3`, `APIGateway`:

```suggestion
- An [AWS user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html) with administrator permissions.
```

5. Copy and paste the corresponding connection string into `env.json`. The following is an example:

    ```json
    {
      "Parameters": {
        "TIDB_HOST": "{gateway-region}.aws.tidbcloud.com",
        "TIDB_PORT": "4000",
        "TIDB_USER": "{prefix}.root",
        "TIDB_PASSWORD": "{password}"
      }
    }
    ```

    Replace the placeholders in `{}` with the values obtained in the connection dialog.

</div>

<div label="TiDB Self-Hosted">

Copy and paste the corresponding connection string into `env.json`. The following is an example:

```json
{
  "Parameters": {
    "TIDB_HOST": "{tidb_server_host}",
    "TIDB_PORT": "4000",
    "TIDB_USER": "root",
    "TIDB_PASSWORD": "{password}"
  }
}
```

Replace the placeholders in `{}` with the values obtained in the **Connect** window.

</div>

</SimpleTab>

### Step 4: Run the code and check the result

1. (Prerequisite) Install the [AWS SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html).

2. Build the bundle:

    ```bash
    npm run build
    ```

> 3. Invoke the sample Lambda function:

Reviewer: **deploy lambda function** should be completed before invoking it or after?

Author: ```suggestion
3. Invoke the sample Lambda function locally:
```

    ```bash
    sam local invoke --env-vars env.json -e events/event.json "tidbHelloWorldFunction"
    ```

4. Check the output in the terminal. If the output is similar to the following, the connection is successful:

    ```bash
    {"statusCode":200,"body":"{\"results\":[{\"Hello World\":\"Hello World\"}]}"}
    ```

> 

Reviewer: ```suggestion

After you confirm that the connection is successful, you can follow the [next section](#deploy-the-aws-lambda-function) to deploy the AWS Lambda Function.

```

```

For more information, refer to [Query data](/develop/dev-guide-get-data-from-single-table.md).

### Update data

The following query adds `50` coins and `50` goods to the `Player` with ID `1`:

```typescript
const [rsh] = await pool.query(
    'UPDATE players SET coins = coins + ?, goods = goods + ? WHERE id = ?;',
    [50, 50, 1]
);
console.log(rsh.affectedRows);
```

For more information, refer to [Update data](/develop/dev-guide-update-data.md).

### Delete data

The following query deletes the `Player` record with ID `1`:

```typescript
const [rsh] = await pool.query('DELETE FROM players WHERE id = ?;', [1]);
console.log(rsh.affectedRows);
```

For more information, refer to [Delete data](/develop/dev-guide-delete-data.md).

## Deploy the AWS Lambda Function

You can deploy the AWS Lambda Function using either the [SAM CLI](#sam-cli-deploymentrecommended) or the [AWS Lambda console](#web-console-deployment).

### SAM CLI deployment(Recommended)

1. ([Prerequisite](#prerequisites)) Install the [AWS SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/install-sam-cli.html).

2. Build the bundle:

    ```bash
    npm run build
    ```

> 3. Update Environment Variables in [`template.yml`](https://github.com/tidb-samples/tidb-aws-lambda-quickstart/blob/main/template.yml)

Reviewer: ```suggestion
3. Update the environment variables in [`template.yml`](https://github.com/tidb-samples/tidb-aws-lambda-quickstart/blob/main/template.yml):
```

    ```yaml
    Environment:
      Variables:
        TIDB_HOST: {tidb_server_host}
        TIDB_PORT: 4000
        TIDB_USER: {prefix}.root
        TIDB_PASSWORD: {password}
    ```

> 4. Set AWS environment variables ([Short-term credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-authentication-short-term.html))

Reviewer: ```suggestion
4. Set AWS environment variables (refer to [Short-term credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-authentication-short-term.html)):
```

    ```bash
    export AWS_ACCESS_KEY_ID={your_access_key_id}
    export AWS_SECRET_ACCESS_KEY={your_secret_access_key}
    export AWS_SESSION_TOKEN={your_session_token}
    ```

> 5. Deploy the AWS Lambda Function

Reviewer: ```suggestion
5. Deploy the AWS Lambda Function:
```

    #        tidbHelloWorldFunction may not have authorization defined, Is this okay? [y/N]: y
    #        Save arguments to configuration file [Y/n]: 
    #        SAM configuration file [samconfig.toml]: 
    #        SAM configuration environment [default]: 

    #        Looking for resources needed for deployment:
    #        Creating the required resources...
    #        Successfully created!
    ```

### Web console deployment

1. Build the bundle:

    ```bash
    npm run build

    # Bundle for AWS Lambda
    # =====================
    # dist/index.zip
    ```

2. Visit the [AWS Lambda console](https://console.aws.amazon.com/lambda/home#/functions).

3. Follow the steps in [Creating a Lambda function](https://docs.aws.amazon.com/lambda/latest/dg/lambda-nodejs.html) to create a Node.js Lambda function.

4. Follow the steps in [Lambda deployment packages](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-package.html#gettingstarted-package-zip) and upload the `dist/index.zip` file.

5. [Copy and configure the corresponding connection string](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html) in Lambda Function.

>     - In the [Functions](https://console.aws.amazon.com/lambda/home#/functions) page of the Lambda console, select the **Configuration** tab, then choose **Environment variables**.
>     - Choose **Edit**.
>     - To add your database access credentials, do the following:

Reviewer: ```suggestion
    1. In the [Functions](https://console.aws.amazon.com/lambda/home#/functions) page of the Lambda console, select the **Configuration** tab, and then choose **Environment variables**.
    2. Choose **Edit**.
    3. To add your database access credentials, do the following:
```

>     - To add your database access credentials, do the following:
>         - Choose **Add environment variable**, then for **Key** enter TIDB_HOST and for **Value** enter the host name.
>         - Choose **Add environment variable**, then for **Key** enter TIDB_PORT and for **Value** enter the port(4000 is default).
>         - Choose **Add environment variable**, then for **Key** enter TIDB_USER and for **Value** enter the user name.
>         - Choose **Add environment variable**, then for **Key** enter TIDB_PASSWORD and for **Value** enter the password you chose when you created your database.

Reviewer 2: ```suggestion
    - To add your database access credentials, do the following:
        - Choose **Add environment variable**, then for **Key** enter `TIDB_HOST` and for **Value** enter the host name.
        - Choose **Add environment variable**, then for **Key** enter `TIDB_PORT` and for **Value** enter the port (4000 is default).
        - Choose **Add environment variable**, then for **Key** enter `TIDB_USER` and for **Value** enter the user name.
        - Choose **Add environment variable**, then for **Key** enter `TIDB_PASSWORD` and for **Value** enter the password you chose when you created your database.
```
